core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,174 @@
-            .collect();
+        let mut public_keys = Vec::new();
+        for role in roles {
+            let role_public_keys = match (role, self.max_faults()) {
+                (Role::Leader, _) => vec![self.sorted_peers[0].public_key.clone()],
+                (Role::ValidatingPeer, 0) => {
+                    if self.sorted_peers.len() > 2 {
+                        vec![self.sorted_peers[1].public_key.clone()]
+                    } else {
+                        vec![]
+                    }
+                }
+                (Role::ProxyTail, 0) => {
+                    if self.sorted_peers.len() == 2 {
+                        vec![self.sorted_peers[1].public_key.clone()]
+                    } else if self.sorted_peers.len() > 2 {
+                        vec![self.sorted_peers[2].public_key.clone()]
+                    } else {
+                        vec![]
+                    }
+                }
+                (Role::ObservingPeer, 0) => {
+                    if self.sorted_peers.len() == 4 {
+                        vec![self.sorted_peers[3].public_key.clone()]
+                    } else {
+                        vec![]
+                    }
+                }
+                (Role::ValidatingPeer, _) => self.sorted_peers
+                    [1..(self.min_votes_for_commit() - 1)]
+                    .iter()
+                    .map(|peer_id| peer_id.public_key.clone())
+                    .collect(),
+                (Role::ProxyTail, _) => vec![self.sorted_peers[self.min_votes_for_commit() - 1]
+                    .public_key
+                    .clone()],
+                (Role::ObservingPeer, _) => self.sorted_peers[self.min_votes_for_commit()..]
+                    .iter()
+                    .map(|peer_id| peer_id.public_key.clone())
+                    .collect(),
+            };
+            public_keys.extend(role_public_keys);
+        }
         signatures
             .into_iter()
             .filter(|signature| public_keys.contains(signature.public_key()))
             .cloned()
             .collect()
     }
-
-    /// Sorted peers that this topology has.
-    pub fn sorted_peers(&self) -> &[PeerId] {
-        &self.sorted_peers
+    /// What role does this peer have in the topology. If it is not in the toplogy
+    /// the function will return `Role::ObservingPeer`.
+    // This lint is a bad suggestion.
+    #[allow(clippy::option_if_let_else)]
+    pub fn role(&self, peer_id: &PeerId) -> Role {
+        match self.sorted_peers.iter().position(|p| p == peer_id) {
+            Some(index) if index == 0 => Role::Leader,
+            Some(index) if index < self.min_votes_for_commit() => Role::ValidatingPeer,
+            Some(index) if index == self.min_votes_for_commit() => Role::ProxyTail,
+            Some(_) => Role::ObservingPeer,
+            None => {
+                trace!(%peer_id, "Peer is not in topology.");
+                Role::ObservingPeer
```

Comment:
maybe we should panic here instead

Reply:
would you elaborate, how can tests rely on this?

Reply:
@mversic Why do you think we should panic here? We ought to be able to recover in this situation.

Reply:
They don't use the consensus code but will create blocks with empty topologies. If we panic here then the tests with empty topologies also panic.

Reply:
> @mversic Why do you think we should panic here? We ought to be able to recover in this situation.

ok, I don't see how it's possible for this situation to occur. I thought it's kinda equivalent to `assert` statements we have around

Reply:
so for empty topologies there is only one peer, the one executing the test? shouldn't `self.sorted_peers` then contain that one peer?

you are free to disregard this comment

Reply:
It's the tests that don't even set up a peer. All they do is create a blockchain. And since the topology exists on the blockheader it get's used. But since there are no peers it is empty.